src/Components/Components/src/NavigationManager.cs

```diff
@@ -274,6 +281,128 @@ protected void NotifyLocationChanged(bool isInterceptedLink)
+        var handlerCount = _locationChangingHandlers.Count;
+        var locationChangingHandlersCopy = ArrayPool<IHandleLocationChanging>.Shared.Rent(handlerCount);
+        _locationChangingHandlers.CopyTo(locationChangingHandlersCopy);
+
+        try
+        {
+            for (var i = 0; i < handlerCount; i++)
+            {
+                await locationChangingHandlersCopy[i].OnLocationChanging(context);
+
+                if (context.IsCanceled || cancellationToken.IsCancellationRequested)
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+        catch (TaskCanceledException ex) when (ex.CancellationToken == cancellationToken)
+        {
+            // If it was our cancellation token, we should gracefully cancel the navigation.
+            return true;
+        }
+        catch (Exception ex)
+        {
+            throw new LocationChangeException("An exception occurred while dispatching a location changing event.", ex);
+        }
+        finally
+        {
+            ArrayPool<IHandleLocationChanging>.Shared.Return(locationChangingHandlersCopy);
+        }
+    }
+
+    /// <summary>
+    /// Sets whether there are active location changing handlers.
+    /// </summary>
+    /// <param name="value">Whether there are active location changing handlers.</param>
+    /// <returns><see langword="true"/> if the change was applied successfully, otherwise <see langword="false"/>.</returns>
+    protected virtual bool SetHasLocationChangingHandlers(bool value)
+        => true;
+
+    /// <summary>
+    /// Invokes <see cref="SetHasLocationChangingHandlers(bool)"/> if the existence of location changing handlers has changed.
+    /// </summary>
+    protected void UpdateHasLocationChangingHandlers()
+    {
+        var hasLocationChangingHandlers = _locationChangingHandlers.Count != 0;
+        if (hasLocationChangingHandlers != _lastUpdatedHasLocationChangingHandlers)
+        {
+            if (SetHasLocationChangingHandlers(hasLocationChangingHandlers))
+            {
+                _lastUpdatedHasLocationChangingHandlers = hasLocationChangingHandlers;
+            }
+        }
+    }
+
+    /// <summary>
+    /// Adds a handler to process incoming navigation events.
+    /// </summary>
+    /// <param name="locationChangingHandler">The handler to process incoming navigation events.</param>
+    public void AddLocationChangingHandler(IHandleLocationChanging locationChangingHandler)
+    {
+        AssertInitialized();
+        _locationChangingHandlers.Add(locationChangingHandler);
+        UpdateHasLocationChangingHandlers();
+    }
+
+    /// <summary>
+    /// Removes a previously-added location changing handler.
+    /// </summary>
+    /// <param name="locationChangingHandler">The handler to remove.</param>
+    /// <returns><see langword="true"/> if the handler was removed, otherwise <see langword="false"/>.</returns>
```

Comment:
What's the use case for this return value? What is a caller meant to do if we return false here?

Reply:
This was mainly to follow the pattern that collections like `List<T>` use, where a `false` return value indicates that the element was not found in the collection and therefore wasn't removed. But for this specific feature, I'm not sure if there's much use in knowing whether the callback you've just removed was even added in the first place. Do you think it's fine to just make the return type `void`?